src/dtype.rs

```diff
@@ -262,22 +288,210 @@ impl PyArrayDescr {
     pub fn shape(&self) -> Vec<usize> {
+        self.as_borrowed().shape()
+    }
+
+    /// Returns true if the type descriptor contains any reference-counted objects in any fields or sub-dtypes.
+    ///
+    /// Equivalent to [`numpy.dtype.hasobject`][dtype-hasobject].
+    ///
+    /// [dtype-hasobject]: https://numpy.org/doc/stable/reference/generated/numpy.dtype.hasobject.html
+    pub fn has_object(&self) -> bool {
+        self.as_borrowed().has_object()
+    }
+
+    /// Returns true if the type descriptor is a struct which maintains field alignment.
+    ///
+    /// This flag is sticky, so when combining multiple structs together, it is preserved
+    /// and produces new dtypes which are also aligned.
+    ///
+    /// Equivalent to [`numpy.dtype.isalignedstruct`][dtype-isalignedstruct].
+    ///
+    /// [dtype-isalignedstruct]: https://numpy.org/doc/stable/reference/generated/numpy.dtype.isalignedstruct.html
+    pub fn is_aligned_struct(&self) -> bool {
+        self.as_borrowed().is_aligned_struct()
+    }
+
+    /// Returns true if the type descriptor is a sub-array.
+    pub fn has_subarray(&self) -> bool {
+        self.as_borrowed().has_subarray()
+    }
+
+    /// Returns true if the type descriptor is a structured type.
+    pub fn has_fields(&self) -> bool {
+        self.as_borrowed().has_fields()
+    }
+
+    /// Returns true if type descriptor byteorder is native, or `None` if not applicable.
+    pub fn is_native_byteorder(&self) -> Option<bool> {
+        self.as_borrowed().is_native_byteorder()
+    }
+
+    /// Returns an ordered list of field names, or `None` if there are no fields.
+    ///
+    /// The names are ordered according to increasing byte offset.
+    ///
+    /// Equivalent to [`numpy.dtype.names`][dtype-names].
+    ///
+    /// [dtype-names]: https://numpy.org/doc/stable/reference/generated/numpy.dtype.names.html
+    pub fn names(&self) -> Option<Vec<&str>> {
+        if !self.has_fields() {
+            return None;
+        }
+        let names = unsafe { Borrowed::from_ptr(self.py(), (*self.as_dtype_ptr()).names) };
+        names.extract().ok()
+    }
+
+    /// Returns the type descriptor and offset of the field with the given name.
+    ///
+    /// This method will return an error if this type descriptor is not structured,
+    /// or if it does not contain a field with a given name.
+    ///
+    /// The list of all names can be found via [`PyArrayDescr::names`].
+    ///
+    /// Equivalent to retrieving a single item from [`numpy.dtype.fields`][dtype-fields].
+    ///
+    /// [dtype-fields]: https://numpy.org/doc/stable/reference/generated/numpy.dtype.fields.html
+    pub fn get_field(&self, name: &str) -> PyResult<(&PyArrayDescr, usize)> {
+        self.as_borrowed()
+            .get_field(name)
+            .map(|(descr, n)| (descr.into_gil_ref(), n))
+    }
+}
+
+/// Implementation of functionality for [`PyArrayDescr`].
+#[doc(alias = "PyArrayDescr")]
+// TODO: seal this trait
```

Comment:
Please do so.

Reply:
Sorry, I thought I did fix this already after the first review of that last PR... Will do 😇.